[WIP] Implement rosservice server in rosserial_server

tongtybj opened the pull request:
Use `topic_tools::ShapeShifter` to create  `proxy` service servers in rosserial server.

Also, during the implementation, I found a very interesting phenomenon.
When the callback function of a `proxy` service server is called,  all process related to `boost::asio` seem to be paused, meaning that no response can be caught from rosserial client unless this service callback function is over. This can be check in c32952f .

So I replace `boost::asio::deadline_timer ros_spin_timer_` with `ros::AsyncSpinner` which has 2 threads. I know this might be a trivial reverting action. But can anyone tell me the motivation to use `boost::asio::deadline_timer ros_spin_timer_` and what [this comment](https://github.com/ros-drivers/rosserial/blob/melodic-devel/rosserial_server/include/rosserial_server/session.h#L156-L157)  means?

tongtybj committed:
Add ServiceServer in rosserial server using topic_tools::ShapeShifter.

tongtybj committed:
User asyncspinner instead of ros timer

tongtybj committed:
Add gtest code for service server in rosserial client

romainreignier commented:
Hi, nice PR!
I have never needed a service server on the `rosserial_client` side but it is nice to see this feature added.

About the async spinner, from what I understand, the current implementation is using the io_service to keep only one thread to simplify the design and avoid synchronization issues.
In your implementation, there is now 3 threads and I do not see any synchronization primitives to protect the resources.

rezeck commented on rosserial_server/include/rosserial_server/topic_handlers.h:
```diff
@@ -184,17 +184,117 @@ class ServiceClient {
+          }
+      }
+
+    response_message = response_message_;
```
Hi, I am getting segfault at this line after I call a custom service at the third time. Any tip how to fix it? I comment this code line and the segfault disappears, but consequently, I do not receive the response message.

tongtybj replied:
@rezeck

Sorry for the very late response. 
~~I have tested  with https://github.com/ros-drivers/rosserial/blob/noetic-devel/rosserial_arduino/src/ros_lib/examples/ServiceServer/ServiceServer.pde, however I can not encounter the problem you pointed out. Could you paste your code to let me test.~~

~~By the way, as @romainreignier pointed out,  there is no  synchronization primitives to protect the resources in this commit, which might be the reason of  your problem. I will add mutex in multiple threads.~~

tongtybj replied:
I have carefully studied this problem again, found there is a fatal bug due to `response_message = response_message_;`
`response_message` which the type is [topic_tools::ShapeShifter](http://docs.ros.org/kinetic/api/topic_tools/html/classtopic__tools_1_1ShapeShifter.html), has a pointer type member `uint8_t * msgBuf`. 

 It seems like the `response_message` will be deleted after this rosservice server callback function, which leads to a invalid access to `msgBuf` in   `response_message_`, in next rosservice response.

I have fixed this problem by commit  3ffc8e4 . Could you please help me to check the latest version of the branch?

mikepurvis committed:
Generate the test client using PYTHON_EXECUTABLE.

mikepurvis committed:
Update pyserial rosdep.

stertingen committed:
Fix warning when using std_msgs/Empty (#482)

stertingen committed:
ros_lib: Removed code duplication in node_handle.h (#483)

* ros_lib: Removed code duplication in node_handle.h

Also, do not generate multiple template instances of
NodeHandle::subscribe()

* Compilefix: call subscribe() with reference, not pointer

daisukes committed:
Use time.sleep instead of rospy.sleep. (#489)

Signed-off-by: Daisuke Sato <daisukes@cmu.edu>

mikepurvis committed:
Use industrial-ci for travis on Noetic.

mikepurvis committed:
Disable rosserial_python test.

mikepurvis committed:
Bump minimum CMake version to 3.7.2 (Melodic).

mikepurvis committed:
Merge pull request #490 from ros-drivers/travis-industrial-ci

Fix Travis for Noetic + Python 3

PetteriAimonen committed:
AVR Float64: fix cases like denormal numbers & inf (#442)

* AVR Float64 conversion: fix special cases like denormal numbers & infinity

* Drop the pthread change.

* Use static_cast in float64 tests

Co-authored-by: Mike Purvis <mike@uwmike.com>

aaraujo11 committed:
Added target ID to support Teensy 4.0 (#460)

mikepurvis committed:
Port 482 and 483 forward from Melodic branch (#492)

stertingen committed:
Do not fail when trying to update ros_lib which already exists (#494)

Co-authored-by: Mike Purvis <mpurvis@clearpath.ai>

mikepurvis committed:
Only initialize embedded python interpreter once. (#491)

stertingen committed:
Use os.path.join for path concatenation (#495)

Co-authored-by: Mike Purvis <mpurvis@clearpath.ai>

rikba committed:
Correctly handle seconds overflow in client time (#497)

asukiaaa committed:
Python 3 and GCC7+ fixes (#508)

* Fixes for new warnings/behaviours in GCC 7+.

* Resolve uninitMemberVar warnings in NodeHandle.

* Port of rosserial_python to py3.

* Fixes and cleanup for rosserial_python py3 port.

* Use nullptr consistently.

* Review fixups.

* Throw from inside the BrokenPipeError.

* Enable python_socket test.

Co-authored-by: Mike Purvis <mpurvis@clearpath.ai>

schnaubelt committed:
Install arduino_python scripts to bin destination (#506)

mikepurvis committed:
Use std::memcpy instead of void casts. (#509)

rikba committed:
Method to compute duration between time stamps (#498)

* Specify duration reference.

* Do not break rosserial_test.

* Implement -operator.

* Add missing bracket.

stertingen committed:
Added support for ChibiOS clients (#493)

* Fix warning when using std_msgs/Empty (#482)

* ros_lib: Removed code duplication in node_handle.h (#483)

* ros_lib: Removed code duplication in node_handle.h

Also, do not generate multiple template instances of
NodeHandle::subscribe()

* Compilefix: call subscribe() with reference, not pointer

* Added support for ChibiOS clients

* Do not fail when trying to update ros_lib which already exists (#494)

Co-authored-by: Mike Purvis <mpurvis@clearpath.ai>

* Use os.path.join for path concatenation

* Bump required CMake version

* Prefer OSAL and ChibiOS time conversion macros if defined

OSAL_I2MS was added to OSAL recently, TIME_I2MS is specifig to
ChibiOS/RT/NIL, so keep the fallback solution

Co-authored-by: Mike Purvis <mpurvis@clearpath.ai>

mikepurvis committed:
Changes

mikepurvis committed:
0.9.0

mikepurvis committed:
Add boost::thread dependency to rosserial_server (#513)

BriceRenaudeau committed:
rosserial_python py3 fix (#514)

mikepurvis committed:
Changes

mikepurvis committed:
0.9.1

mikepurvis committed:
Fix header mismatch in changelog.

tongtybj committed:
Resolve the double free or corruption problem in the rosservice callback function.

tongtybj commented:
@romainreignier 

> In your implementation, there is now 3 threads and I do not see any synchronization primitives to protect the resources.

I think the only shared resource that should be protected is the write (`async_write`) process, since the ros callback functions and several dealine_timers do the write process from different threads.
 
 So I add a mutex to  lock the following part:

https://github.com/ros-drivers/rosserial/blob/50a94094e334e7c99d41e6138a5bb5dcdd02b7ab/rosserial_server/include/rosserial_server/session.h#L267-L270

rafaol committed:
Changed C++ STL include and code into original C versions (#525)

mcamurri committed:
Python3 import fix for rosserial_arduino (#554)

Tobias-Fischer committed:
Add python3-dev as build depend (#544)

VasilyF committed:
Fix CMP0038 warning in arduino-cmake (#553)

Nurgak committed:
Gracefully handle socket server end. (#550)

lopsided98 committed:
Use shutil.copyfile() instead of shutil.copy() (#530)

On my distribution, rosserial's files are installed readonly, and shutil.copy() preserves this when copying them. The causes make_libraries to fail the next time it runs as it tries to overwrite the readonly files. shutil.copyfile() does not preserve the permissions, preventing this problem.

tongtybj committed:
Correct module import rule for Python3 (#521)

altineller committed:
Adds a shutdown hook for sending txStopRequest (#551)

Co-authored-by: Mike Purvis <mpurvis@clearpath.ai>

mikepurvis committed:
Changes

mikepurvis committed:
0.9.2

nnarain committed:
Sync if multiple unrecognised topics are encountered (#563)

asukiaaa committed:
Remove ESP flags for ArduinoTcpHardware to use hardware serial (#559)

alonblade committed:
node_handle: spinOnce: add two new return codes (#556)

Adds two new return codes (backward compatible change):

SPIN_TX_STOP_REQUESTED
SPIN_TIME_RECV

They are returned as the name suggests to differentiate two states in
which currently SPIN_OK would have been returned.

This allows the caller to further use the return, i.e. issuing a log
message (that is always sent irrespective of connected status since ID <
100).

Signed-off-by: Alon Levy <alon.levy@bladeranger.com>

TobinHall committed:
Fixing message size issue. (#516)

amalpavithran committed:
Add optional private parameter config to Python server (#562)

* Made tcp_port a local parameter

The tcp_port parameter being a global parameter prevents the launch of multiple serial nodes on different port numbers using a launch file

* Added Fallback & Modified fork_server parameter

* Removed unnecessary defaults & added comments

tongtybj committed:
Refactor the wating timeout for service server

tongtybj committed:
Change the level of verbose for sync lost

tongtybj committed:
Add arg for serial.launch

tongtybj committed:
Switch the service reponse and reuqest md5sum check

tongtybj committed:
Separate subscriber callbacks and service server callbacks into two different AsyncSpinner to run them in different threads.

tongtybj committed:
Send ID_TX_STOP on shutdown.

tongtybj committed:
Compatible usage of boost asio dispath for different boost version

mjseppa committed:
Replace busy wait loop with condition variables

mjseppa committed:
Match sync timeout with rosserial_python

mjseppa committed:
Fix subscriber typo

tongtybj committed:
[rosserial_test] remove unecessary dependency in CMake

tongtybj committed:
Merge pull request #5 from mjseppa/rosservice_server

Rosservice server condition variable wait and timing fix for rosserial_python

sugikazu75 committed:
fix deserialization for nested array

sugikazu75 committed:
remove unnecessary semi-colon for in make_library

tongtybj committed:
Merge pull request #6 from sugikazu75/rosservice_server

[rosserial_clieant] fix deserialization for nested array
